Magnetic field geometry builder DD4hep migration (part 1)

cvuosalo opened the pull request:
#### PR description:

Migrate the magnetic field geometry builder to DD4hep, except for reading from the DB. Code has been migrated as-is with very minimal changes. The existing code could be rewritten and revised, but that task is not within the scope of this PR. The principles of this PR:

    Continue to support the old DD, along with the new DD4hep
    Keep changes to old code to the bare minimum
    Share as much code between the old and new DD methods as possible

Migration of reading the magnetic field from the DB with DD4hep will be done in a subsequent PR.

Note that the need for the pseudo-trapezoid shape is uncertain at this time. Its final status will be resolved in a later PR.

This PR should not change the behavior of current workflows.

This PR is a replacement for #27674.

#### PR validation:

A unit test is included in this PR, and it thoroughly tests the geometry builder. This test was run for the old DD and the new DD4hep version, and the results were the same. Regression tests for the old DD and new DD4hep also produced matching results. Other units tests and matrix tests were also successful.

No backport needed.

cvuosalo committed:
Condensed commit for mf geom builder migration to DD4hep

cvuosalo committed:
Condensed commit for mf geom builder migration to DD4hep

cvuosalo committed:
Remove unneeded CLHEP

cvuosalo committed:
Formatted again

cvuosalo committed:
Formatted again

cvuosalo committed:
Revert unnecessary change to MagGeometryAnalyzer.cc

cvuosalo committed:
Revert unnecessary change to MagGeometryAnalyzer.cc

cvuosalo committed:
Clarify comment and remove unused data member

cvuosalo committed:
Clarify comment and remove unused data member

cvuosalo committed:
Addressing first part of Slava's comments. Currently not compiling

cvuosalo committed:
Addressing first part of Slava's comments. Currently not compiling

cvuosalo committed:
Push since IB has expired -- need to rebase. Addressing PR comments, but not compiling yet.

cvuosalo committed:
Compile errors fixed

cvuosalo committed:
Update for latest version of external DD4hep

cvuosalo committed:
Code formatting fixes

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-27886/11664

- This PR adds an extra 152KB to repository

- Found files with invalid states:
  - MagneticField/GeomBuilder/plugins/dd4hep/volumeHandle.h:
    - Added: 71594334871360510fa34ed947cd7ae61bf056ed, d7ef3514316b7d9622a505cfe180297bf1839657
    - Deleted: daf6836511b597f303416f87e1c8c2a7ccc05659, 173748c01a90198ce2b6b315fd4c780e4cd610be
  - MagneticField/GeomBuilder/plugins/dd4hep/MagGeoBuilder.h:
    - Added: 71594334871360510fa34ed947cd7ae61bf056ed, d7ef3514316b7d9622a505cfe180297bf1839657
    - Deleted: daf6836511b597f303416f87e1c8c2a7ccc05659, 173748c01a90198ce2b6b315fd4c780e4cd610be
  - MagneticField/GeomBuilder/plugins/dd4hep/VolBasedMagFieldESProducerNewDD.cc:
    - Added: 71594334871360510fa34ed947cd7ae61bf056ed, d7ef3514316b7d9622a505cfe180297bf1839657
    - Modified: 7d8f7ff05872e5505561ec9e0b581aecafc29f09, 55c6e0e1593e031e169fb80ab9f343c6e0e95051
    - Deleted: daf6836511b597f303416f87e1c8c2a7ccc05659, 173748c01a90198ce2b6b315fd4c780e4cd610be
  - MagneticField/GeomBuilder/plugins/dd4hep/MagGeoBuilder.cc:
    - Added: 71594334871360510fa34ed947cd7ae61bf056ed, d7ef3514316b7d9622a505cfe180297bf1839657
    - Modified: d15b540043fcf96aab9217f60aaf1a7420a3349e, 4bff86e378b5a2ce16d3002a80e705e67e122223
    - Deleted: daf6836511b597f303416f87e1c8c2a7ccc05659, 173748c01a90198ce2b6b315fd4c780e4cd610be
  - MagneticField/GeomBuilder/plugins/dd4hep/volumeHandle.cc:
    - Added: 71594334871360510fa34ed947cd7ae61bf056ed, d7ef3514316b7d9622a505cfe180297bf1839657
    - Modified: 7d8f7ff05872e5505561ec9e0b581aecafc29f09, 55c6e0e1593e031e169fb80ab9f343c6e0e95051
    - Deleted: daf6836511b597f303416f87e1c8c2a7ccc05659, 173748c01a90198ce2b6b315fd4c780e4cd610be


- There are other open Pull requests which might conflict with changes you have proposed:
  - File  DetectorDescription/DDCMS/src/DDFilteredView.cc  modified in PR(s): #27883

cmsbuild commented:
A new Pull Request was created by  @cvuosalo (Carl Vuosalo) for master.

It involves the following packages:

DetectorDescription/DDCMS
MagneticField/GeomBuilder
MagneticField/Records
MagneticField/VolumeBasedEngine


@perrotta, @civanch, @Dr15Jones, @cvuosalo, @ianna, @mdhildreth, @cmsbuild, @slava77 can you please review it and eventually sign? Thanks.
@namapane, @vargasa this is something you requested to watch as well.
@davidlange6, @slava77, @fabiocos you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: e6cb3afbe6b6a69f1120303610432b2b8a749e8c
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a1ed12/2226/summary.html

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a1ed12/2226/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 34
 - DQMHistoTests: Total histograms compared: 2939508
 - DQMHistoTests: Total failures: 1
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2939166
 - DQMHistoTests: Total skipped: 341
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 33 files compared)
 - Checked 145 log files, 15 edm output root files, 34 DQM output files

ianna commented:
+1

ianna commented on DetectorDescription/DDCMS/src/DDFilteredView.cc:
```diff
@@ -2,6 +2,7 @@
 #include "FWCore/MessageLogger/interface/MessageLogger.h"
 #include "DD4hep/Detector.h"
+#include "DD4hep/Shapes.h"
 #include <TGeoBBox.h>
```
it's just a better version of the same thing. It's ok to go in.

slava77 commented:
from https://github.com/cms-sw/cmssw/pull/27886#issuecomment-525852991
> @gartung @smuzaffar @Dr15Jones it seems like the static analyzer is not running with `EDM_ML_DEBUG`

Please clarify/confirm if this is the expected behavior now (I seem to recall that it was turned to use ```EDM_ML_DEBUG``` at some point a year or two ago)

gartung commented:
@slava77 The move to the latest clang version most likely broke something in macro expansion. Looking into this requires time I don't have at the moment.

smuzaffar commented:
@slava77 , only for IBs we are runing static analyze with  EDM_ML_DEBUG ( https://github.com/cms-sw/cms-bot/blob/master/run_class_dumper.sh#L16 ). I do not remeber enabling it for PR tests. Looks like we should, any objections @Dr15Jones ?

Dr15Jones commented:
No objection

smuzaffar commented:
@slava77 , cms-sw/cms-bot#1197 PR enables EDM_ML_DEBUG for tests. Should we re-start tests for this PR along with cms-sw/cms-bot#1197 ?

slava77 commented:
> >  Should we re-start tests for this PR along with [cms-sw/cms-bot#1197](https://github.com/cms-sw/cms-bot/pull/1197)


yes, please

smuzaffar commented:
please test with cms-sw/cms-bot#1197

cmsbuild commented:
The tests are being triggered in jenkins.
Tested with other pull request(s) cms-sw/cms-bot#1197

cmsbuild commented:
+1
Tested at: e6cb3afbe6b6a69f1120303610432b2b8a749e8c
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-89b5f9/2262/summary.html

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-89b5f9/2262/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 4 differences found in the comparisons
 - DQMHistoTests: Total files compared: 34
 - DQMHistoTests: Total histograms compared: 2939508
 - DQMHistoTests: Total failures: 2
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2939165
 - DQMHistoTests: Total skipped: 341
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 33 files compared)
 - Checked 145 log files, 15 edm output root files, 34 DQM output files

cvuosalo commented:
The tests show one minor issue. There is a harmless, unnecessary `const_cast` left behind in the old code:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-89b5f9/2262/llvm-analysis/report-a40b8c.html#EndPath

In `const_cast<Surface *>(surfaces[i].get())`, `surfaces` is now non-const so the cast is not needed, but it is harmless. Can this issue be fixed in the next PR? I don't know if modifying and pushing this branch will work. Or is it safe to edit the source code file in Git? I've never tried that.

slava77 reviewed (commented):
most of the comments are apparently to the parts modified since the last review

namapane commented on MagneticField/GeomBuilder/src/BaseVolumeHandle.cc:
```diff
@@ -0,0 +1,220 @@
+  return false;
+}
+
+bool BaseVolumeHandle::setSurface(Surface &s1, Sides which_side) {
```
I suppose the cosnt chane is related to https://github.com/cms-sw/cmssw/pull/27674#discussion_r316906238.
Apart for this, for what I remember this method was intended to be used in a further development of the MF geometry representation, where neighboring volumes would share surface, making it possible for a propagator to navigate without the need of global volume search. As you know well, an easier solution was eventually adopted. I am ok with removing the method if you insist, although in the spirit of this PR I would suggest to stick to DD4hep modifications here.

cvuosalo replied:
`setSurface` is not currently used, but I would like to leave to Nicola the decision of whether or not to delete it.

namapane commented on MagneticField/GeomBuilder/src/BaseVolumeHandle.h:
```diff
@@ -0,0 +1,219 @@
+    virtual DDSolidShape shape() const = 0;
+
+  protected:
+    typedef ReferenceCountingPointer<Surface> RCPS;
```
Same as above, cf: https://github.com/cms-sw/cmssw/pull/27674#discussion_r316906238

cvuosalo replied:
As Nicola mentioned, my previous comment explained the issue. Here's some more detail. The `VolumeSide` constructor takes a non-const pointer to `Surface`. `VolumeSide` is defined in `"MagneticField/VolumeGeometry/interface/VolumeSide.h`, outside the scope of this PR. In the `volumeHandle::sides()` method, `Surface` pointers from the `surfaces` member are passed to the `VolumeSide` constructor. In the old code, `surfaces` was defined to contain const pointers, and a const_cast was used to remove the constness. To avoid the static analzyzer complaint about the const_cast, I made `surfaces` non-const by changing the `RCPS` typedef.
Since the pointers contained in `surfaces` are passed to `VolumeSide`, it seems to make sense to make `surfaces` non-const. However, we could go back to the original approach, which is to keep `surfaces` const until the pointers need to be passed to `VolumeSide`, at which point they must be const_cast to take away the constness.

Let me know which approach we should follow in this PR.

slava77 replied:
> Let me know which approach we should follow in this PR.

from what I found in github cmssw, the preferred solution to me is to make VolumeSide use ConstReferenceCountingPointer and also a ```const Surface*``` in the interface.
It seems like this flexibility for VolumeSide to let the surface be modified was never needed and it leads to awkward code patterns.

I'm fine with simply reverting to the old state using const_cast (it seems to me anyways more acceptable than the current state)

cvuosalo replied:
I am restoring the old way, including the const_cast. I will leave `VolumeSide` revision to Nicola.

namapane commented on MagneticField/GeomBuilder/src/DD4hep_MagGeoBuilder.cc:
```diff
@@ -0,0 +1,609 @@
+  }    // end for
+  iunique = ptrs.size();
+
+  LogTrace("MagGeoBuilder") << "    volumes   " << ivolumes << "    surfaces  " << isurfaces << "    assigned  "
```
I must say that I would strongly prefer the formatting not to be changed. These printouts are just needed when debugging a new geometry, which is something I do myself when Slava Kl. prepares a new one. The debug output is truly gigantic, and having to sort out changes in formatting will be a hassle. I have been diffing the outputs to figure out  issues in the past, for example.

cvuosalo replied:
Sorry for being sloppy with the debug messages. I will restore all their formatting.

namapane commented on MagneticField/GeomBuilder/src/DD4hep_volumeHandle.cc:
```diff
@@ -0,0 +1,208 @@
+// #include "Utilities/Configuration/interface/Architecture.h"
```
Relic from ancient past... can go

cvuosalo replied:
Deleted.

namapane commented on MagneticField/GeomBuilder/src/DD4hep_volumeHandle.cc:
```diff
@@ -0,0 +1,208 @@
+      continue;
+
+    RCPS s = surfaces[i].get();
+    result.push_back(VolumeSide(s, GlobalFace(i), surfaces[i]->side(center_, 0.3)));
```
I'm not sure of what you mean, but keep in mind that Surfaces are reference counted so passing them around with normal counters can potentially result in them being deleted when you don't expect it. So please, do not replace RCPS's with normal pointers anyhwere.

cvuosalo replied:
I would like to leave `VolumeSide` for another PR.

namapane commented on MagneticField/GeomBuilder/src/buildBox.icc:
```diff
@@ -69,28 +70,26 @@ void MagGeoBuilderFromDDD::volumeHandle::buildBox(const DDExpandedView &fv) {
+      << pos_inner << " " << pos_inner.perp() << " " << pos_inner.phi() << " pos_zplus    " << pos_zplus << " "
+      << pos_zplus.perp() << " " << pos_zplus.phi() << " pos_zminus   " << pos_zminus << " " << pos_zminus.perp() << " "
+      << pos_zminus.phi() << " pos_phiplus  " << pos_phiplus << " " << pos_phiplus.perp() << " " << pos_phiplus.phi()
+      << " pos_phiminus " << pos_phiminus << " " << pos_phiminus.perp() << " " << pos_phiminus.phi();
```
Please do not use LogDebug but only LogTrace anyhwere you replace couts. LogDebug fills the gigantic debug output with headers that make it  impossible to use the debug output for the purpose I needed to code it. I already think it is going to be a nightmare to fix all these debug messages to sort them out when I will need to look at them the next time...

cvuosalo replied:
I restored and cleaned up the debug message formatting.
I also changed all `LogDebug` to `LogTrace` in all files in this PR.

namapane commented on MagneticField/GeomBuilder/src/buildPseudoTrap.icc:
```diff
@@ -81,42 +81,41 @@ void MagGeoBuilderFromDDD::volumeHandle::buildPseudoTrap(const DDExpandedView &f
+      LogDebug("magneticfield::volumeHandle") << setprecision(10);
+      LogDebug("magneticfield::volumeHandle") << "*** WARNING: Cylinder surface not centered on beam axis " << rcheck
+                                              << " " << fR << " " << std::abs(rcheck - fR);
+      LogDebug("magneticfield::volumeHandle") << setprecision(6);
```
LogTrace, NO LogDebug please...

cvuosalo replied:
Fixed.

namapane commented on MagneticField/GeomBuilder/src/DD4hep_MagGeoBuilder.cc:
```diff
@@ -0,0 +1,609 @@
+    LogTrace("MagGeoBuilder") << "*** BARREL ********************************************"
+                              << " Number of different volumes   = " << bVolCount
+                              << " Number of interpolators built = " << bInterpolators.size()
+                              << " Number of MagBLayers built    = " << mBLayers_.size();
```
because this if triggers the testInside() method (see below). It is a consistency test that is run only in debug mode; it should stay within the if.

cvuosalo replied:
Debug message formatting restored. I moved the `if (debug_)` to only enclose the `testInside()` so the `LogTrace` message is controlled independently of the `debug_` flag.

cvuosalo commented:
Trying a test push to this branch. I hope it doesn't wreck the PR.

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-27886/11711

- This PR adds an extra 152KB to repository

- Found files with invalid states:
  - MagneticField/GeomBuilder/plugins/dd4hep/volumeHandle.h:
    - Added: 71594334871360510fa34ed947cd7ae61bf056ed, d7ef3514316b7d9622a505cfe180297bf1839657
    - Deleted: daf6836511b597f303416f87e1c8c2a7ccc05659, 173748c01a90198ce2b6b315fd4c780e4cd610be
  - MagneticField/GeomBuilder/plugins/dd4hep/MagGeoBuilder.h:
    - Added: 71594334871360510fa34ed947cd7ae61bf056ed, d7ef3514316b7d9622a505cfe180297bf1839657
    - Deleted: daf6836511b597f303416f87e1c8c2a7ccc05659, 173748c01a90198ce2b6b315fd4c780e4cd610be
  - MagneticField/GeomBuilder/plugins/dd4hep/VolBasedMagFieldESProducerNewDD.cc:
    - Added: 71594334871360510fa34ed947cd7ae61bf056ed, d7ef3514316b7d9622a505cfe180297bf1839657
    - Modified: 7d8f7ff05872e5505561ec9e0b581aecafc29f09, 55c6e0e1593e031e169fb80ab9f343c6e0e95051
    - Deleted: daf6836511b597f303416f87e1c8c2a7ccc05659, 173748c01a90198ce2b6b315fd4c780e4cd610be
  - MagneticField/GeomBuilder/plugins/dd4hep/MagGeoBuilder.cc:
    - Added: 71594334871360510fa34ed947cd7ae61bf056ed, d7ef3514316b7d9622a505cfe180297bf1839657
    - Modified: d15b540043fcf96aab9217f60aaf1a7420a3349e, 4bff86e378b5a2ce16d3002a80e705e67e122223
    - Deleted: daf6836511b597f303416f87e1c8c2a7ccc05659, 173748c01a90198ce2b6b315fd4c780e4cd610be
  - MagneticField/GeomBuilder/plugins/dd4hep/volumeHandle.cc:
    - Added: 71594334871360510fa34ed947cd7ae61bf056ed, d7ef3514316b7d9622a505cfe180297bf1839657
    - Modified: 7d8f7ff05872e5505561ec9e0b581aecafc29f09, 55c6e0e1593e031e169fb80ab9f343c6e0e95051
    - Deleted: daf6836511b597f303416f87e1c8c2a7ccc05659, 173748c01a90198ce2b6b315fd4c780e4cd610be


- There are other open Pull requests which might conflict with changes you have proposed:
  - File  DetectorDescription/DDCMS/src/DDFilteredView.cc  modified in PR(s): #27883, #27907

cmsbuild commented:
Pull request #27886 was updated. @perrotta, @civanch, @Dr15Jones, @cvuosalo, @ianna, @mdhildreth, @cmsbuild, @slava77 can you please check and sign again.

cvuosalo commented on MagneticField/GeomBuilder/src/DD4hep_MagGeoBuilder.cc:
```diff
@@ -0,0 +1,609 @@
+    LogTrace("MagGeoBuilder") << "*** ENDCAP ********************************************"
+                              << " Number of different volumes   = " << eVolCount
+                              << " Number of interpolators built = " << eInterpolators.size()
+                              << " Number of MagESector built    = " << mESectors_.size();
```
Fixed as above.

slava77 commented:
@cvuosalo 
there is also a conflict in  DetectorDescription/DDCMS/src/DDFilteredView.cc 
please check and fix.
Thank you.

cvuosalo commented on MagneticField/GeomBuilder/src/buildBox.icc:
```diff
@@ -69,28 +70,26 @@ void MagGeoBuilderFromDDD::volumeHandle::buildBox(const DDExpandedView &fv) {
-    cout << " halfX: " << halfX << " halfY: " << halfY << " halfZ: " << halfZ << endl
-         << "RN            " << theRN << endl;
+  LogDebug("magneticfield::volumeHandle")
+      << " halfX: " << halfX << " halfY: " << halfY << " halfZ: " << halfZ << " RN            " << theRN;
```
It is a short message. I kept it on one line and made it more readable.

cvuosalo commented on MagneticField/GeomBuilder/src/buildBox.icc:
```diff
@@ -102,15 +101,13 @@ void MagGeoBuilderFromDDD::volumeHandle::buildBox(const DDExpandedView &fv) {
-  }
+  LogDebug("magneticfield::volumeHandle") << "rot_R " << surfaces[outer]->toGlobal(LocalVector(0., 0., 1.)) << " rot_Z "
+                                          << surfaces[zplus]->toGlobal(LocalVector(0., 0., 1.)) << " rot_phi "
+                                          << surfaces[phiplus]->toGlobal(LocalVector(0., 0., 1.));
```
Debug message formatting is restored.

cvuosalo commented on MagneticField/GeomBuilder/src/buildPseudoTrap.icc:
```diff
@@ -81,42 +81,41 @@ void MagGeoBuilderFromDDD::volumeHandle::buildPseudoTrap(const DDExpandedView &f
-    cout << "z_phiminus " << z_phiminus << " " << z_phiminus.phi() << endl;
-  }
+  LogDebug("magneticfield::volumeHandle") << "z_phiplus " << z_phiplus << " " << z_phiplus.phi();
+  LogDebug("magneticfield::volumeHandle") << "z_phiminus " << z_phiminus << " " << z_phiminus.phi();
```
Debug message combined into one `LogTrace`.

cvuosalo commented on MagneticField/GeomBuilder/src/buildTrap.icc:
```diff
@@ -103,10 +103,8 @@ void MagGeoBuilderFromDDD::volumeHandle::buildTrap(const DDExpandedView &fv) {
-    cout << "y_phiminus " << y_phiminus << endl;
-  }
+  LogDebug("magneticfield::volumeHandle") << "y_phiplus " << y_phiplus;
+  LogDebug("magneticfield::volumeHandle") << "y_phiminus " << y_phiminus;
```
Debug messages combined into one `LogTrace`.

cvuosalo commented on MagneticField/GeomBuilder/src/buildTrap.icc:
```diff
@@ -135,16 +133,17 @@ void MagGeoBuilderFromDDD::volumeHandle::buildTrap(const DDExpandedView &fv) {
-
-    if (fabs(surfaces[phiminus]->position().phi() - thePhiMin) > 1e-5) {
-      cout << " Trapez. phiMin = " << thePhiMin << endl;
+  if (debug) {
```
I moved the `if (debug)` to enclose only the debug calculation.

cvuosalo commented on MagneticField/GeomBuilder/src/buildTrap.icc:
```diff
@@ -135,16 +133,17 @@ void MagGeoBuilderFromDDD::volumeHandle::buildTrap(const DDExpandedView &fv) {
+        << surfaces[phiminus]->toGlobal(LocalVector(0., 0., 1.)).phi();
+
+    if (std::abs(surfaces[phiminus]->position().phi() - thePhiMin) > 1e-5) {
+      LogDebug("magneticfield::volumeHandle") << " Trapez. phiMin = " << thePhiMin;
```
`LogDebug` converted to `LogTrace` in all the files.

cvuosalo committed:
Fix second round of PR comments

cvuosalo commented on MagneticField/GeomBuilder/src/volumeHandle.cc:
```diff
@@ -422,29 +236,18 @@ std::vector<VolumeSide> MagGeoBuilderFromDDD::volumeHandle::sides() const {
+}  // namespace
 
 #include "MagneticField/GeomBuilder/src/buildBox.icc"
 #include "MagneticField/GeomBuilder/src/buildTrap.icc"
```
Removed unneeded directory name from includes.

cvuosalo commented on MagneticField/GeomBuilder/src/DD4hep_volumeHandle.h:
```diff
@@ -0,0 +1,61 @@
+  typedef const char* ShapeType;
+  class DDFilteredView;
+
+  class volumeHandle : public magneticfield::BaseVolumeHandle {
```
@slava77: As you suggest, I will put the new DD4hep classes in the `cms::magneticfield` namespace. I hope it doesn't create confusion with the plain `magneticfield` namespace.

cvuosalo replied:
@slava77: I had to change plans. I would like to keep the `magneticfield` namespace for use in some of the old files. The `cms` namespace is used for DD4hep-related names. Mixing `cms::magneticfield` and `magneticfield` did not seem to compile, so I created the `cms::dd4hepmagfield` namespace for the DD4hep-related names that are used for the magnetic field. I hope that is all right for a name.

cvuosalo commented on MagneticField/GeomBuilder/src/buildPseudoTrap.icc:
```diff
@@ -81,42 +81,41 @@ void MagGeoBuilderFromDDD::volumeHandle::buildPseudoTrap(const DDExpandedView &f
-      cout << setprecision(6);
+  if (debug) {
+    if (std::abs(rcheck - fR) > 100. * epsilon) {  //FIXME!
+      LogDebug("magneticfield::volumeHandle") << setprecision(10);
```
I doubt `setprecision` works for `LogTrace`. I deleted them.

namapane replied:
I start being irritated about these modifications to debug messages. This PR is about migration to DD4hep, not about cleanup of debug messages, but we are now mostly talking about and changing forth and back the latter. I do need these messages and I do need them to issue messages formatted just as they were coded. If LogDebug does not support setprecision can we leave the cout as it was? I'm really unhappy about having to reformat the messages and possibly even to recode them as couts here and there in my private version (if I need formatting that LogTrace does not support) when I will need them. @slava77, I know that is the policy, but I doubt anybody else has ever run this code with debug on in the past 15 years, so I really wonder what is the purpose of messing up the PR with all these changes to debug messages.

slava77 replied:
I did not push for switching to Log\*, however, once it was done, I was motivated to see it applied uniformly.
A large fraction of my review was in order to preserve the old formatting.

cvuosalo replied:
For the files being added or revised to work with DD4hep, I replaced `cout` with `LogTrace` because there is a general recommendation to not use `cout` in code that is not in the `test` directory. The Log utilities are recommended instead. I did not change `cout` in the old code because I wanted to avoid major revisions and updating of that code.
I will try to ensure that the formatting of the debug messages is not changed by this PR.

cvuosalo replied:
I was able to restore the `setprecision` statement. Note that this file is not used with the DD4hep version because we don't expect there will be any pseudo-trapezoids.

cvuosalo committed:
Part 2 of fixing second round of comments

cvuosalo commented on MagneticField/GeomBuilder/src/DD4hep_MagGeoBuilder.h:
```diff
@@ -0,0 +1,94 @@
+#ifndef MagneticField_GeomBuilder_DD4hepMagGeoBuilder_h
+#define MagneticField_GeomBuilder_DD4hepMagGeoBuilder_h
```
Fixed spelling of ifndef guard in `DD4hep_MagGeoBuilder.h`.

cvuosalo committed:
Namespace adjustments

cvuosalo committed:
Code formatting

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-27886/11780

- This PR adds an extra 160KB to repository

- Found files with invalid states:
  - MagneticField/GeomBuilder/plugins/dd4hep/volumeHandle.h:
    - Added: 71594334871360510fa34ed947cd7ae61bf056ed, d7ef3514316b7d9622a505cfe180297bf1839657
    - Deleted: daf6836511b597f303416f87e1c8c2a7ccc05659, 173748c01a90198ce2b6b315fd4c780e4cd610be
  - MagneticField/GeomBuilder/plugins/dd4hep/MagGeoBuilder.h:
    - Added: 71594334871360510fa34ed947cd7ae61bf056ed, d7ef3514316b7d9622a505cfe180297bf1839657
    - Deleted: daf6836511b597f303416f87e1c8c2a7ccc05659, 173748c01a90198ce2b6b315fd4c780e4cd610be
  - MagneticField/GeomBuilder/plugins/dd4hep/VolBasedMagFieldESProducerNewDD.cc:
    - Added: 71594334871360510fa34ed947cd7ae61bf056ed, d7ef3514316b7d9622a505cfe180297bf1839657
    - Modified: 7d8f7ff05872e5505561ec9e0b581aecafc29f09, 55c6e0e1593e031e169fb80ab9f343c6e0e95051
    - Deleted: daf6836511b597f303416f87e1c8c2a7ccc05659, 173748c01a90198ce2b6b315fd4c780e4cd610be
  - MagneticField/GeomBuilder/plugins/dd4hep/MagGeoBuilder.cc:
    - Added: 71594334871360510fa34ed947cd7ae61bf056ed, d7ef3514316b7d9622a505cfe180297bf1839657
    - Modified: d15b540043fcf96aab9217f60aaf1a7420a3349e, 4bff86e378b5a2ce16d3002a80e705e67e122223
    - Deleted: daf6836511b597f303416f87e1c8c2a7ccc05659, 173748c01a90198ce2b6b315fd4c780e4cd610be
  - MagneticField/GeomBuilder/plugins/dd4hep/volumeHandle.cc:
    - Added: 71594334871360510fa34ed947cd7ae61bf056ed, d7ef3514316b7d9622a505cfe180297bf1839657
    - Modified: 7d8f7ff05872e5505561ec9e0b581aecafc29f09, 55c6e0e1593e031e169fb80ab9f343c6e0e95051
    - Deleted: daf6836511b597f303416f87e1c8c2a7ccc05659, 173748c01a90198ce2b6b315fd4c780e4cd610be


- There are other open Pull requests which might conflict with changes you have proposed:
  - File  DetectorDescription/DDCMS/interface/DDFilteredView.h  modified in PR(s): #27926
  - File  DetectorDescription/DDCMS/src/DDFilteredView.cc  modified in PR(s): #27926


Code check has found code style and quality issues which could be resolved by applying following patch(s)
- **code-checks**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-27886/11780/code-checks.patch
e.g. `curl https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-27886/11780/code-checks.patch | patch -p1`
You can also run `scram build code-checks` to apply code checks directly

cvuosalo commented:
Still in process. Working on the rebasing and checking debug message formatting.

cmsbuild commented:
The code-checks are being triggered in jenkins.

cvuosalo commented:
Minor fixes still coming

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-27886/11787

- This PR adds an extra 216KB to repository

- Found files with invalid states:
  - MagneticField/GeomBuilder/plugins/dd4hep/volumeHandle.h:
    - Added: 957d80d086fb0b5622f59510ed3b08429459e353, 6aa8543b01656d69cf8fcf455de29de9f0bf3ac1
    - Deleted: 786571247f9faa4601aad17d53170b4186645eef, b65032059b281e48f101ebee19e1b21339018998
  - MagneticField/GeomBuilder/plugins/dd4hep/MagGeoBuilder.h:
    - Added: 957d80d086fb0b5622f59510ed3b08429459e353, 6aa8543b01656d69cf8fcf455de29de9f0bf3ac1
    - Deleted: 786571247f9faa4601aad17d53170b4186645eef, b65032059b281e48f101ebee19e1b21339018998
  - MagneticField/GeomBuilder/plugins/dd4hep/VolBasedMagFieldESProducerNewDD.cc:
    - Added: 957d80d086fb0b5622f59510ed3b08429459e353, 6aa8543b01656d69cf8fcf455de29de9f0bf3ac1
    - Modified: eaaed4b685055dba1bc7a64783da7c3aefb24b15, 78d71aff0afc8ebee1e4444acebe69a0a93c1336
    - Deleted: 786571247f9faa4601aad17d53170b4186645eef, b65032059b281e48f101ebee19e1b21339018998
  - MagneticField/GeomBuilder/plugins/dd4hep/MagGeoBuilder.cc:
    - Added: 957d80d086fb0b5622f59510ed3b08429459e353, 6aa8543b01656d69cf8fcf455de29de9f0bf3ac1
    - Modified: e6d1e35a04f7d8713ab890e9914a2e2433d99672, 9cc0070ba23adf0a8c23d43a49f0651e5fb790db
    - Deleted: 786571247f9faa4601aad17d53170b4186645eef, b65032059b281e48f101ebee19e1b21339018998
  - MagneticField/GeomBuilder/plugins/dd4hep/volumeHandle.cc:
    - Added: 957d80d086fb0b5622f59510ed3b08429459e353, 6aa8543b01656d69cf8fcf455de29de9f0bf3ac1
    - Modified: eaaed4b685055dba1bc7a64783da7c3aefb24b15, 78d71aff0afc8ebee1e4444acebe69a0a93c1336
    - Deleted: 786571247f9faa4601aad17d53170b4186645eef, b65032059b281e48f101ebee19e1b21339018998

cmsbuild commented:
Pull request #27886 was updated. @perrotta, @civanch, @Dr15Jones, @cvuosalo, @ianna, @mdhildreth, @cmsbuild, @slava77 can you please check and sign again.

cvuosalo committed:
Fix formatting of debug messages

cmsbuild commented:
The code-checks are being triggered in jenkins.

cvuosalo commented on MagneticField/GeomBuilder/BuildFile.xml:
```diff
@@ -11,6 +12,7 @@
 <use   name="Utilities/BinningTools"/>
 <use   name="CondFormats/MFObjects"/>
 <use   name="clhep"/>
+  <use name="dd4hep"/>
```
Fixed. dd4hep is not needed.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-27886/11802

- This PR adds an extra 216KB to repository

- Found files with invalid states:
  - MagneticField/GeomBuilder/plugins/dd4hep/volumeHandle.h:
    - Added: 957d80d086fb0b5622f59510ed3b08429459e353, 6aa8543b01656d69cf8fcf455de29de9f0bf3ac1
    - Deleted: 786571247f9faa4601aad17d53170b4186645eef, b65032059b281e48f101ebee19e1b21339018998
  - MagneticField/GeomBuilder/plugins/dd4hep/MagGeoBuilder.h:
    - Added: 957d80d086fb0b5622f59510ed3b08429459e353, 6aa8543b01656d69cf8fcf455de29de9f0bf3ac1
    - Deleted: 786571247f9faa4601aad17d53170b4186645eef, b65032059b281e48f101ebee19e1b21339018998
  - MagneticField/GeomBuilder/plugins/dd4hep/VolBasedMagFieldESProducerNewDD.cc:
    - Added: 957d80d086fb0b5622f59510ed3b08429459e353, 6aa8543b01656d69cf8fcf455de29de9f0bf3ac1
    - Modified: eaaed4b685055dba1bc7a64783da7c3aefb24b15, 78d71aff0afc8ebee1e4444acebe69a0a93c1336
    - Deleted: 786571247f9faa4601aad17d53170b4186645eef, b65032059b281e48f101ebee19e1b21339018998
  - MagneticField/GeomBuilder/plugins/dd4hep/MagGeoBuilder.cc:
    - Added: 957d80d086fb0b5622f59510ed3b08429459e353, 6aa8543b01656d69cf8fcf455de29de9f0bf3ac1
    - Modified: e6d1e35a04f7d8713ab890e9914a2e2433d99672, 9cc0070ba23adf0a8c23d43a49f0651e5fb790db
    - Deleted: 786571247f9faa4601aad17d53170b4186645eef, b65032059b281e48f101ebee19e1b21339018998
  - MagneticField/GeomBuilder/plugins/dd4hep/volumeHandle.cc:
    - Added: 957d80d086fb0b5622f59510ed3b08429459e353, 6aa8543b01656d69cf8fcf455de29de9f0bf3ac1
    - Modified: eaaed4b685055dba1bc7a64783da7c3aefb24b15, 78d71aff0afc8ebee1e4444acebe69a0a93c1336
    - Deleted: 786571247f9faa4601aad17d53170b4186645eef, b65032059b281e48f101ebee19e1b21339018998


- There are other open Pull requests which might conflict with changes you have proposed:
  - File  DetectorDescription/DDCMS/data/cms-mf-geometry.xml  modified in PR(s): #27946

cmsbuild commented:
Pull request #27886 was updated. @perrotta, @civanch, @Dr15Jones, @cvuosalo, @ianna, @mdhildreth, @cmsbuild, @slava77 can you please check and sign again.

cvuosalo commented:
@cmsbuild please test with #27933

cmsbuild commented:
The tests are being triggered in jenkins.
Tested with other pull request(s) cms-sw/cmssw#27933

slava77 commented:
> @cmsbuild please test with #27933

is this the right companion PR for testing?
#27933 was rejected with a note that #27946 should be used instead

cvuosalo commented:
@slava77: I think I addressed all your comments with the latest changes, except for those I asked to defer to a later PR. These changes are mostly about fixing the formatting of the debug messages. I found that `std::endl` is ignored by `LogTrace`. To get new-lines into the messages, I had to use "\n". I created a constant called `newln` that is equal to "\n", and I replaced ` std::endl` with `newln` to get nicely formatted debug messages. In studying the debug messages, I found even some of the old ones that weren't quite lined up properly, and I added a space here or there to make them all look nice. I will post examples of the debug messages.
Unfortunately, the automatic code formatting changes the arrangement of the code for the debug messages. This doesn't change the actual output, but it does cause difficulty in looking at the code and seeing that the debug message will come out properly formatted.

cvuosalo commented:
@slava77: Concerning the tests, I am getting to that. I wanted to put to rest the previous issues before raising a new one. The goal of the current test is just to show the code is OK at this point before we go onto the next step.

cvuosalo commented:
Here are examples of the debug messages from the current version of the PR to show they are properly formatted now:
```
RN            377.931
pos_outer     (386.79,68.7777,-421.655)  392.858 0.175977
pos_inner     (369.071,68.7777,-421.655)  375.425 0.18424
pos_zplus     (377.931,68.7777,-181.3)  384.138 0.180015
pos_zminus    (377.931,68.7777,-662.01)  384.138 0.180015
pos_phiplus   (377.931,137.555,-421.655)  402.185 0.349066
pos_phiminus  (377.931,3.05176e-05,-421.655)  377.931 8.07491e-08
y_phiplus   (0.939693,0.342018,6.5359e-17) 
y_phiminus  (1,2.67208e-06,1.14126e-16) 
rot_R     (1,-5.05322e-16,1.14127e-16) 
rot_Z     (1.14127e-16,-1.22465e-16,-1) 
rot_phi+  (0.342018,-0.939694,1.54113e-16)  phi -1.22173
rot_phi-  (2.67208e-06,-1,1.22465e-16)  phi -1.57079
 RMin =  369.071
 RMax =  386.79
Summary: V_1138 1 shape = DDSolidShape index:3, name: Trapezoid trasl  (377.931,68.7777,-421.655)  R 384.138 phi 0.180015 magFile  Material= materials:Air isIron= 0 masterSector= 1
 Orientation of surfaces:
  0:negativeSide
  1:positiveSide
  2:positiveSide
  3:negativeSide
  4:positiveSide
  5:negativeSide
 Vol R and Z values determine barrel or endcap. R = 384.138, Z = -421.655
 (Barrel)
Building interpolator from 1138 copyno 1 at  (377.931,68.7777,-421.655)  phi: 0.0573006 pi,  file: s01/grid.1138.bin master: 1
Grid has 3 dimensions  number of nodes is 3 5 53
Volume:1138 : Number of grid points outside the MagVolume: 0/795
Name: V_1138
Rotation matrix components (1st 3) = -0.5, 0.866025, 1.13875e-16
Refplane pos   (292.909,248.529,-421.655) 
Building trapezoid surfaces...: 
x1 67.1654
x2 70.39
x3 67.1654
x4 70.39
y1 8.85965
y2 8.85965
theta  0
phi    0
halfZ  240.355
alpha1 0.180015
alpha2 0.180015
```
```
zhalf      313
rIn        216
rOut       720.045
startPhi   0
deltaPhi   0.261799
cutAtStart 695.51
cutAtDelta 720.045
cutInside  0
Actual Center at:  (451.856,59.488,-2087)  R 455.755 phi 0.1309
RN            455.755
pos_zplus     (451.856,59.488,-1774)  455.755 0.1309
pos_zminus    (451.856,59.488,-2400)  455.755 0.1309
pos_phiplus   (440.226,117.958,-2087)  455.755 0.261799
pos_phiminus  (455.755,-5.58139e-14,-2087)  455.755 -1.22465e-16
y_phiplus   (0.965926,0.258819,1.02444e-16) 
y_phiminus  (1,-1.22465e-16,1.22465e-16) 
rot_Z     (-1.22465e-16,6.12323e-17,1) 
rot_phi+  (-0.258819,0.965926,-9.08421e-17)  phi 1.8326
rot_phi-  (1.22465e-16,1,-6.12323e-17)  phi 1.5708
pos_Rplane     (695.51,-8.51754e-14,-2087)  695.51 -1.22465e-16
rot_R          (1,3.33533e-07,1.22465e-16)  phi 3.33533e-07
cyl radius     216
 RMin =  216
 RMax =  720.045
```

cvuosalo commented:
@slava77: The next issue with this PR comes about from a new PR that was created and merged over the past weekend. Up to now, this code, both old and new DD4hep, used a single `DDSolidShape` enum class. The new development is the addition of another `DDSolidShape`, almost the same as the old one, but not quite the same. I wanted to wrap up and merge this PR and then address the new `DDSolidShape` in a separate PR. However, Yana insisted that too much other development depended on the new `DDSolidShape` so that it couldn't be reverted. That means we have to deal with it in this PR.

Slava, I want to handle this issue in the way that is easiest and most convenient for your review. You have the opportunity now to review the latest changes to see that your comments were addressed without the confounding effects of a new set of changes. I can wait until you have checked the current changes before pushing new ones.

On the other hand, if you'd prefer to do only one review and review all the changes at once, I can push the new changes as soon as I can, which won't be until Monday or Tuesday. The new changes will affect about seven files in this PR.

Please let me know which approach would work best for you.

cmsbuild commented:
+1
Tested at: 91ccfe9bbd69baea24bbd7e568161361d3caf1d2
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-ee80c7/2418/summary.html

cmsbuild commented:
Comparison job queued.

slava77 commented:
> That means we have to deal with it in this PR.

@cvuosalo 
I'm not sure I understand this well enough.
It looks like this PR compiled and ran OK.
If your planned commits will be separated (no force-push/rebase) on top what is now here, I can review based on a commit range via web github.

I do not plan to look at the details in this PR until Monday.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-ee80c7/2418/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 34
 - DQMHistoTests: Total histograms compared: 2955796
 - DQMHistoTests: Total failures: 1
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2955454
 - DQMHistoTests: Total skipped: 341
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 33 files compared)
 - Checked 145 log files, 15 edm output root files, 34 DQM output files

ianna commented:
> > @cmsbuild please test with #27933
> 
> is this the right companion PR for testing?
> #27933 was rejected with a note that #27946 should be used instead

It was a mistake to bind theses PRs. The #27933 will not go in that means this one cannot be merged.

cvuosalo commented:
@ianna: I am not suggesting that this PR be merged in its current state. I wanted to show that the PR is clean and working up to now. Then next week I will add the `DDSolidShape` changes.

ianna commented:
@smuzaffar - is it possible to revert the "requires-external" state for this PR? thanks.

fabiocos commented:
@ianna @cvuosalo if the companion PR used for test is not really needed the flag in itself is not a problem

cvuosalo committed:
Add conversion of second, new DDSolidShape related to DD4hep

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-27886/11845

- This PR adds an extra 156KB to repository

- Found files with invalid states:
  - MagneticField/GeomBuilder/plugins/dd4hep/volumeHandle.h:
    - Added: 957d80d086fb0b5622f59510ed3b08429459e353, 6aa8543b01656d69cf8fcf455de29de9f0bf3ac1
    - Deleted: 786571247f9faa4601aad17d53170b4186645eef, b65032059b281e48f101ebee19e1b21339018998
  - MagneticField/GeomBuilder/plugins/dd4hep/MagGeoBuilder.h:
    - Added: 957d80d086fb0b5622f59510ed3b08429459e353, 6aa8543b01656d69cf8fcf455de29de9f0bf3ac1
    - Deleted: 786571247f9faa4601aad17d53170b4186645eef, b65032059b281e48f101ebee19e1b21339018998
  - MagneticField/GeomBuilder/plugins/dd4hep/VolBasedMagFieldESProducerNewDD.cc:
    - Added: 957d80d086fb0b5622f59510ed3b08429459e353, 6aa8543b01656d69cf8fcf455de29de9f0bf3ac1
    - Modified: eaaed4b685055dba1bc7a64783da7c3aefb24b15, 78d71aff0afc8ebee1e4444acebe69a0a93c1336
    - Deleted: 786571247f9faa4601aad17d53170b4186645eef, b65032059b281e48f101ebee19e1b21339018998
  - MagneticField/GeomBuilder/plugins/dd4hep/MagGeoBuilder.cc:
    - Added: 957d80d086fb0b5622f59510ed3b08429459e353, 6aa8543b01656d69cf8fcf455de29de9f0bf3ac1
    - Modified: e6d1e35a04f7d8713ab890e9914a2e2433d99672, 9cc0070ba23adf0a8c23d43a49f0651e5fb790db
    - Deleted: 786571247f9faa4601aad17d53170b4186645eef, b65032059b281e48f101ebee19e1b21339018998
  - MagneticField/GeomBuilder/plugins/dd4hep/volumeHandle.cc:
    - Added: 957d80d086fb0b5622f59510ed3b08429459e353, 6aa8543b01656d69cf8fcf455de29de9f0bf3ac1
    - Modified: eaaed4b685055dba1bc7a64783da7c3aefb24b15, 78d71aff0afc8ebee1e4444acebe69a0a93c1336
    - Deleted: 786571247f9faa4601aad17d53170b4186645eef, b65032059b281e48f101ebee19e1b21339018998

cmsbuild commented:
Pull request #27886 was updated. @perrotta, @civanch, @Dr15Jones, @cvuosalo, @ianna, @mdhildreth, @cmsbuild, @slava77 can you please check and sign again.

cvuosalo commented:
@slava77: I have now changed this PR to be compatible with #27907 and #27946, both of which are now merged. Those two PRs introduced a new `cms::DDSolidShape` that could have collided with the old `DDSolidShape` used in this PR.
To prevent collision, I removed `using namespace cms;` statements and referred to all objects in the `cms` namespace explicitly, e.g. `cms::DDFilteredView` (with one small exception, see below). I also changed the namespace of the new DD4hep classes in this PR from `cms` to `magneticfield`.
The new `cms::DDSolidShape` only appears implicitly as the return value of the `cms::dd::getCurrentShape()` function, which gets immediately converted to the old `DDSolidShape` by the `cms::DDFilteredView::legacyShape()` method. Throughout the rest of the code only the old `DDSolidShape` is used.

I hope that the issues with this PR are now addressed.

cvuosalo commented:
Here's a technical note about special usage of the `cms` namespace. The `MagneticField/GeomBuilder/src/build*.icc` files are included in both the old DD and new DD4hep code. They instantiate objects like `DDBox`, which have two versions for the old and new DD. To use the new version of these objects, I needed to put a `using namespace cms::dd;` statement immediately before the `#include` lines for these files. These files don't use `DDSolidShape` or other conflicting objects so the `using namespace` statement does not cause any collisions.

cvuosalo commented:
With the last change, this PR does not require #27933. Is it safe to run the comparison test now even though this PR was tested with #27933 previously?
We might also want to wait on the test because the required PR #27946 was only merged three hours ago.

ianna commented on MagneticField/GeomBuilder/src/bLayer.h:
```diff
@@ -1,55 +1,58 @@
+    handles theVolumes;            // pointer to all volumes in this layer
+
+    mutable MagBLayer* mlayer;
+    const bool debug;
```
here is a result of that static member change.

ianna commented on MagneticField/GeomBuilder/src/bRod.h:
```diff
@@ -8,30 +8,32 @@
+    std::vector<bSlab> slabs;
+    handles volumes;  // pointers to all volumes in the rod
+    mutable MagBRod* mrod;
+    bool debug;  // Allow assignment from other bRod objects
```
and here is a result of that static member change. The size of bool is 1 byte, but how many bRod and other objects are there?

namapane replied:
The most ubiquitous are volumeHandle which are of the order of ~2k; the total number of concerned instances is probably less than 3k. Consider that these are temporary objects, and are deleted after the map construction is done.

ianna commented on MagneticField/GeomBuilder/src/bRod.cc:
```diff
@@ -6,19 +6,21 @@
  *  \author N. Amapane - INFN Torino
  */
 
-#include "MagneticField/GeomBuilder/src/bRod.h"
```
please, remove:
```// #include "Utilities/Configuration/interface/Architecture.h"```

cvuosalo replied:
Fixed.

ianna commented on MagneticField/GeomBuilder/src/bSector.cc:
```diff
@@ -6,35 +6,36 @@
  *  \author N. Amapane - INFN Torino
  */
 
-#include "MagneticField/GeomBuilder/src/bSector.h"
```
please, remove:
```// #include "Utilities/Configuration/interface/Architecture.h"```

cvuosalo replied:
Fixed.

ianna commented on MagneticField/VolumeBasedEngine/src/MagGeometry.cc:
```diff
@@ -118,8 +129,12 @@ MagVolume const* MagGeometry::findVolume1(const GlobalPoint& gp, double toleranc
   } else {  // Endcaps
     for (vector<MagVolume6Faces const*>::const_iterator v = theEVolumes.begin(); v != theEVolumes.end(); ++v) {
```
```for( auto const& v : theEVolumes)```

namapane replied:
The code was written well before this syntax made its way in the C++ standard (even before "auto"). While I am (obviously) not unhappy if the code gets updated to the latest standard, I think that is a bit off the scope of the current PR.

cvuosalo replied:
This file was only changed because it was causing crashes and excess error messages while I was debugging. I put in some safety checks, but more revision of this file is totally unrelated to DD4hep migration.

ianna commented on MagneticField/VolumeBasedEngine/src/MagGeometry.cc:
```diff
@@ -63,12 +66,15 @@ MagGeometry::MagGeometry(int geomVersion,
 MagGeometry::~MagGeometry() {
-  delete theBarrelBinFinder;
-  delete theEndcapBinFinder;
+  if (theBarrelBinFinder != nullptr)
```
why not ```std::unique_ptr```?

namapane replied:
std::unique_ptr was not existing in C++ when the code was written. Overhauling the package to use unique_ptr consistently (ie not just here, but in all similar cases, which are quite a few) is obviously possible, but I think that is beyond the scope of the current PR, and personally I don't think it's worth the effort - "if it ain't broke..."

cvuosalo replied:
I agree with Nicola.

ianna commented on MagneticField/VolumeBasedEngine/src/MagGeometry.cc:
```diff
@@ -63,12 +66,15 @@ MagGeometry::MagGeometry(int geomVersion,
+    delete theEndcapBinFinder;
 
   for (vector<MagBLayer const*>::const_iterator ilay = theBLayers.begin(); ilay != theBLayers.end(); ++ilay) {
     delete (*ilay);
```
```for( auto const& play : theBLayers)```

ianna reviewed (commented):
All the comments except for the debug static member change can be addressed in a later PR.

ianna commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

namapane commented on MagneticField/GeomBuilder/src/MagGeoBuilderFromDDD.h:
```diff
@@ -123,6 +98,6 @@ class MagGeoBuilderFromDDD {
   const magneticfield::TableFileMap* theGridFiles;  // Non-owned pointer assumed to be valid until build() is called
 
-  static bool debug;
+  const bool debug;
```
My understanding is that Carl has decided to move away from a single static used everywhere to a model where the flag is percolated. I am ok with that as far as design is concerned.

cvuosalo replied:
For the debugging strategy, moving to using `#ifdef EDM_ML_DEBUG` instead of a debug flag could even be considered. It would go along with `LogTrace` and would remove any performance penalty of storing debug flags and of processing `if (debug)` statements but would not allow turning on debugging in a program running in production.

Nicola has a more complete view of the debugging needs of this code, so I would like to leave the debugging design to him. I think I should stop fiddling with it.

cmsbuild commented:
+1
Tested at: 0f38562dc980f64b224fdd3d93ea85abc06eb48a
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a1ed12/2476/summary.html

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a1ed12/2476/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 34
 - DQMHistoTests: Total histograms compared: 2957224
 - DQMHistoTests: Total failures: 1
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2956882
 - DQMHistoTests: Total skipped: 341
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 33 files compared)
 - Checked 145 log files, 15 edm output root files, 34 DQM output files

ianna commented:
+1

slava77 reviewed (commented):
it looks like all my comments here are related to the most recent updates.
We are very close though.

cvuosalo commented on MagneticField/GeomBuilder/src/buildBox.icc:
```diff
@@ -69,28 +70,27 @@ void MagGeoBuilderFromDDD::volumeHandle::buildBox(const DDExpandedView &fv) {
-         << "pos_phiminus " << pos_phiminus << " " << pos_phiminus.perp() << " " << pos_phiminus.phi() << endl;
+  LogTrace("magneticfield::volumeHandle")
+      << "pos_outer    " << pos_outer << " " << pos_outer.perp() << " " << pos_outer.phi() << newln << "pos_inner    "
+      << pos_inner << " " << pos_inner.perp() << " " << pos_inner.phi() << newln << "pos_zplus    " << pos_zplus << " "
```
I carefully lined up these statements so they would look nice when reading the code. Then clang-format reformatted them the "correct" CMS way. Sorry, I don't think there is anything I can do. The code formatting occurs automatically in every PR, as I understand it.

slava77 replied:
strange.
So, it doesn't realign if there is an endl (old code), but reshuffles for other cases.

cvuosalo commented on MagneticField/GeomBuilder/src/buildPseudoTrap.icc:
```diff
@@ -10,42 +11,37 @@
-         << "atMinusZ  " << atMinusZ << endl;
+  LogTrace("magneticfield::volumeHandle")
+      << "halfZ     " << halfZ << newln << "x1        " << x1 << newln << "x2        " << x2 << newln << "y1        "
+      << y1 << newln << "y2        " << y2 << newln << "radius    " << radius << newln << "atMinusZ  " << atMinusZ;
```
clang-format did this, and all the later ones, too.

cvuosalo commented on MagneticField/GeomBuilder/src/bLayer.cc:
```diff
@@ -6,29 +6,33 @@
  *  \author N. Amapane - INFN Torino
  */
 
-#include "MagneticField/GeomBuilder/src/bLayer.h"
```
Fixed.

cvuosalo commented on MagneticField/GeomBuilder/src/MagGeoBuilderFromDDD.h:
```diff
@@ -27,6 +26,8 @@ namespace magneticfield {
   class VolumeBasedMagneticFieldESProducerFromDB;
   class AutoMagneticFieldESProducer;
+  class BaseVolumeHandle;  // Needs to be public to share code with DD4hep
+  typedef std::vector<BaseVolumeHandle*> handles;
```
Fixed.

cvuosalo commented on MagneticField/GeomBuilder/src/BaseVolumeHandle.h:
```diff
@@ -0,0 +1,225 @@
+}  // namespace magneticfield
+
+namespace {
+  const char* const newln = "\n";
```
How about if I put `newln` into the `magneticfield` namespace instead? I am also adding `inline` to tell the compiler to make only one copy of the constant.
I created this `newln` constant to look like and take the place of `endl`. It is purely cosmetic. I think it looks nicer than `"\n"`. I don't really care. I could change to using `"\n"` explicitly, but it wouldn't look as nice to my eye.

cvuosalo committed:
Move newln into magneticfield namespace

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-27886/11891

- This PR adds an extra 144KB to repository

- Found files with invalid states:
  - MagneticField/GeomBuilder/plugins/dd4hep/volumeHandle.h:
    - Added: 957d80d086fb0b5622f59510ed3b08429459e353, 6aa8543b01656d69cf8fcf455de29de9f0bf3ac1
    - Deleted: 786571247f9faa4601aad17d53170b4186645eef, b65032059b281e48f101ebee19e1b21339018998
  - MagneticField/GeomBuilder/plugins/dd4hep/MagGeoBuilder.h:
    - Added: 957d80d086fb0b5622f59510ed3b08429459e353, 6aa8543b01656d69cf8fcf455de29de9f0bf3ac1
    - Deleted: 786571247f9faa4601aad17d53170b4186645eef, b65032059b281e48f101ebee19e1b21339018998
  - MagneticField/GeomBuilder/plugins/dd4hep/VolBasedMagFieldESProducerNewDD.cc:
    - Added: 957d80d086fb0b5622f59510ed3b08429459e353, 6aa8543b01656d69cf8fcf455de29de9f0bf3ac1
    - Modified: eaaed4b685055dba1bc7a64783da7c3aefb24b15, 78d71aff0afc8ebee1e4444acebe69a0a93c1336
    - Deleted: 786571247f9faa4601aad17d53170b4186645eef, b65032059b281e48f101ebee19e1b21339018998
  - MagneticField/GeomBuilder/plugins/dd4hep/MagGeoBuilder.cc:
    - Added: 957d80d086fb0b5622f59510ed3b08429459e353, 6aa8543b01656d69cf8fcf455de29de9f0bf3ac1
    - Modified: e6d1e35a04f7d8713ab890e9914a2e2433d99672, 9cc0070ba23adf0a8c23d43a49f0651e5fb790db
    - Deleted: 786571247f9faa4601aad17d53170b4186645eef, b65032059b281e48f101ebee19e1b21339018998
  - MagneticField/GeomBuilder/plugins/dd4hep/volumeHandle.cc:
    - Added: 957d80d086fb0b5622f59510ed3b08429459e353, 6aa8543b01656d69cf8fcf455de29de9f0bf3ac1
    - Modified: eaaed4b685055dba1bc7a64783da7c3aefb24b15, 78d71aff0afc8ebee1e4444acebe69a0a93c1336
    - Deleted: 786571247f9faa4601aad17d53170b4186645eef, b65032059b281e48f101ebee19e1b21339018998


Code check has found code style and quality issues which could be resolved by applying following patch(s)
- **code-format**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-27886/11891/code-format.patch
e.g. `curl https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-27886/11891/code-format.patch | patch -p1`
You can also run `scram build code-format` to apply code format directly

cvuosalo committed:
One blank line removed by code-format

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-27886/11892

- This PR adds an extra 144KB to repository

- Found files with invalid states:
  - MagneticField/GeomBuilder/plugins/dd4hep/volumeHandle.h:
    - Added: 957d80d086fb0b5622f59510ed3b08429459e353, 6aa8543b01656d69cf8fcf455de29de9f0bf3ac1
    - Deleted: 786571247f9faa4601aad17d53170b4186645eef, b65032059b281e48f101ebee19e1b21339018998
  - MagneticField/GeomBuilder/plugins/dd4hep/MagGeoBuilder.h:
    - Added: 957d80d086fb0b5622f59510ed3b08429459e353, 6aa8543b01656d69cf8fcf455de29de9f0bf3ac1
    - Deleted: 786571247f9faa4601aad17d53170b4186645eef, b65032059b281e48f101ebee19e1b21339018998
  - MagneticField/GeomBuilder/plugins/dd4hep/VolBasedMagFieldESProducerNewDD.cc:
    - Added: 957d80d086fb0b5622f59510ed3b08429459e353, 6aa8543b01656d69cf8fcf455de29de9f0bf3ac1
    - Modified: eaaed4b685055dba1bc7a64783da7c3aefb24b15, 78d71aff0afc8ebee1e4444acebe69a0a93c1336
    - Deleted: 786571247f9faa4601aad17d53170b4186645eef, b65032059b281e48f101ebee19e1b21339018998
  - MagneticField/GeomBuilder/plugins/dd4hep/MagGeoBuilder.cc:
    - Added: 957d80d086fb0b5622f59510ed3b08429459e353, 6aa8543b01656d69cf8fcf455de29de9f0bf3ac1
    - Modified: e6d1e35a04f7d8713ab890e9914a2e2433d99672, 9cc0070ba23adf0a8c23d43a49f0651e5fb790db
    - Deleted: 786571247f9faa4601aad17d53170b4186645eef, b65032059b281e48f101ebee19e1b21339018998
  - MagneticField/GeomBuilder/plugins/dd4hep/volumeHandle.cc:
    - Added: 957d80d086fb0b5622f59510ed3b08429459e353, 6aa8543b01656d69cf8fcf455de29de9f0bf3ac1
    - Modified: eaaed4b685055dba1bc7a64783da7c3aefb24b15, 78d71aff0afc8ebee1e4444acebe69a0a93c1336
    - Deleted: 786571247f9faa4601aad17d53170b4186645eef, b65032059b281e48f101ebee19e1b21339018998

cmsbuild commented:
Pull request #27886 was updated. @perrotta, @civanch, @Dr15Jones, @cvuosalo, @ianna, @mdhildreth, @cmsbuild, @slava77 can you please check and sign again.

cvuosalo commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: b20b6263f2d5bd19c51912f43578f0a422af539b
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a1ed12/2516/summary.html

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a1ed12/2516/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 34
 - DQMHistoTests: Total histograms compared: 2957224
 - DQMHistoTests: Total failures: 1
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2956882
 - DQMHistoTests: Total skipped: 341
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 33 files compared)
 - Checked 145 log files, 15 edm output root files, 34 DQM output files

slava77 commented:
+1

for https://github.com/cms-sw/cmssw/pull/27886	b20b626 
- code changes are in line with the PR description and mostly in line with the follow up review
- jenkins tests pass and comparisons with the baseline show no differences

cvuosalo commented:
+1

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will now be reviewed by the release team before it's merged. @davidlange6, @slava77, @smuzaffar, @fabiocos (and backports should be raised in the release meeting by the corresponding L2)

fabiocos commented:
+1

fabiocos commented:
@cvuosalo in the PR description you say that the test configurations available have been used to verify that the magnetic field description is the same between old and new tool. But for what I can see in my check their output format is quite different and not trivially comparable out-of-the-box. How was this check done? I see that with some output manipulation the final part of the output is comparable, although here and there there are differences...

namapane commented:
@fabiocos The output messages you are referring to are intended for debugging, not for validation.
For validation we have a point-by-point regression that compares with pre-stored values of the field. This validates both the actual interpolation, and the underlying geometry (gaps, association between volumes and interpolators, etc).
One caveat is that the current new-DD producer works from XML files; building from the database (which is the default) still goes through the old DDD.

cvuosalo commented:
@fabiocos: Nicola provided a very good answer. The two test procedures are very thorough, checking literally millions of points to ensure that the magnetic field geometry was built correctly by the code. The code in this PR passed the tests just like the old code.
